### PR TITLE
Always write the cache (unless cache_dir is /dev/null)

### LINKED ISSUE #3045: Always write the `-i` cache
I've noticed that many people initially run plain `mypy` and then start using `mypy -i` but find that the first such run is not any faster than a plain run. This is understandable because it needs to warm the cache first.

We could alleviate this issue by just always writing the cache (perhaps allowing this feature to be turned off for special scenarios but having it on by default). This might cost a little bit of extra run time -- IIRC I once measured the cost of writing the cache and it was in the 5-10% range.

### PR DESCRIPTION
**Pull request by gvanrossum**
Fixes #3045

(How to test? I tested manually and it seems to work. It's also pretty straightforward.)

### CODE REVIEW

File: mypy/build.py [hunk 1]
```diff
@@ -1523,26 +1528,26 @@ def valid_references(self) -> Set[str]:
...
-                list(self.dependencies), list(self.suppressed), list(self.child_modules),
-                dep_prios, self.interface_hash,
-                self.manager)
-            if new_interface_hash == self.interface_hash:
-                self.manager.log("Cached module {} has same interface".format(self.id))
-            else:
-                self.manager.log("Cached module {} has changed interface".format(self.id))
-                self.mark_interface_stale()
-                self.interface_hash = new_interface_hash
+        if not self.path or self.options.cache_dir == os.devnull:
```
Comment (ambv) [inline mypy/build.py]: Alright, we can still override writing the cache by setting /dev/null.

👍

File: mypy/build.py [hunk 2]
```diff
@@ -891,7 +891,12 @@ def write_cache(id: str, path: str, tree: MypyFile,
         manager.trace("Interface for {} has changed".format(id))
```
Comment (ambv) [inline mypy/build.py]: Shouldn't the `replace` in :890 be only done if there's no OSError in :895?

File: mypy/build.py [hunk 3]
```diff
@@ -891,7 +891,12 @@ def write_cache(id: str, path: str, tree: MypyFile,
...
 
     # Obtain and set up metadata
-    st = manager.get_stat(path)  # TODO: Handle errors
+    try:
+        st = manager.get_stat(path)
+    except OSError:
+        manager.log("Cannot write {}".format(path))
```
Comment (ambv) [inline mypy/build.py]: 1. We should try to delete the nonced file here (and the data file; it's invalidated by the mismatched `data_mtime` anyway).
2. Would also be good to at least stringify the OSError to get an idea why the stat failed.
3. The error message is misleading. It's a source stat issue, not a cache write issue.

**Comment by gvanrossum** [comment]
OK, I think I've made all the changes you suggested. I moved the get_stat() call up earlier so we don't even write anything if we can't stat() the source; and I also delete the cache files in that case (not sure if it's even needed now).

**Comment by gvanrossum** [comment]
A downside is that it slows down regular runs somewhat. (5-10%? I have to collect more serious data.)

File: mypy/build.py [hunk 4]
```diff
@@ -876,6 +876,20 @@ def write_cache(id: str, path: str, tree: MypyFile,
         data_str = json.dumps(data, sort_keys=True)
     interface_hash = compute_hash(data_str)
 
+    # Obtain and set up metadata
+    try:
+        st = manager.get_stat(path)
+    except OSError as err:
+        manager.log("Cannot get stat for {}: %s".format(path, err))
```
Comment (ilevkivskyi) [inline mypy/build.py]: Something is wrong with formatting here, this does not fail, but ``%s`` is never replaced (at least in Python 3.4 I just tried).

Comment (gvanrossum) [inline mypy/build.py]: Good eye, I am still not in the habit of using `{}` for `format()`...

File: mypy/build.py [hunk 5]
```diff
@@ -860,8 +860,10 @@ def write_cache(id: str, path: str, tree: MypyFile,
 
     # Make sure directory for cache files exists
     parent = os.path.dirname(data_json)
-    if not os.path.isdir(parent):
+    try:
         os.makedirs(parent)
+    except os.error:
+        pass  # Assume it already exists.
```
Comment (JelleZijlstra) [inline mypy/build.py]: You can just call `os.makedirs(parent, exist_ok=True)` instead (as of 3.2, https://docs.python.org/3/library/os.html#os.makedirs).

Comment (gvanrossum) [inline mypy/build.py]: Oh thanks! Why this isn't the default behavior I'll never understand (it's Eric Raymond's fault :-).

**Pull request merged**